quesma/ingest/ingest_validator.go

```diff
@@ -169,63 +97,66 @@ func validateNumericType(columnType string, incomingValueType string, value inte
 	return false
 }
 
-func validateValueAgainstType(fieldName string, value interface{}, targetColumnType clickhouse.Type) types.JSON {
-	// Validate Array() types by recursing on the inner type:
-	if compoundType, isCompound := targetColumnType.(clickhouse.CompoundType); isCompound && compoundType.Name == "Array" {
-		if valueAsArray, isArray := value.([]interface{}); isArray && len(valueAsArray) > 0 {
-			innerTypesAreCompatible := true
-			innerTypeName := getTypeName(valueAsArray[0])
+func validateValueAgainstType(fieldName string, value interface{}, columnType clickhouse.Type) (isValid bool) {
+	switch columnType := columnType.(type) {
+	case clickhouse.BaseType:
+		incomingValueType, err := clickhouse.NewType(value, fieldName)
+		if err != nil {
+			return false
+		}
+
+		columnTypeName := removeLowCardinality(columnType.Name)
+
+		if isNumericType(columnTypeName) {
+			if incomingValueType, isBaseType := incomingValueType.(clickhouse.BaseType); isBaseType && validateNumericType(columnTypeName, incomingValueType.Name, value) {
+				// Numeric types match!
+				return true
+			}
+		}
 
-			// Make sure that all elements of the array have the same type
-			for _, e := range valueAsArray {
-				eTypeName := getTypeName(e)
-				// Check if the type names are exactly the same. However, if it's a numeric type, perform more advanced
-				// validation (to handle an array like [3.5, 4, 4.5]).
-				if isNumericType(eTypeName) && isNumericType(innerTypeName) {
-					if !validateNumericType(innerTypeName, eTypeName, e) {
-						innerTypesAreCompatible = false
-						break
+		if incomingValueType, isBaseType := incomingValueType.(clickhouse.BaseType); isBaseType && incomingValueType.Name == columnTypeName {
+			// Types match exactly!
+			return true
+		}
+
+		return false
+	case clickhouse.MultiValueType:
+		if columnType.Name == "Tuple" {
+			if value, isMap := value.(map[string]interface{}); isMap {
+				for key, elem := range value {
+					subtype := columnType.GetColumn(key)
+					if subtype == nil {
+						return false
 					}
-				} else {
-					if getTypeName(e) != innerTypeName {
-						innerTypesAreCompatible = false
-						break
+					if !validateValueAgainstType(fmt.Sprintf("%s.%s", fieldName, key), elem, subtype.Type) {
+						return false
 					}
 				}
+				return true
 			}
+		} else {
+			logger.Error().Msgf("MultiValueType validation is not yet supported for type: %v", columnType)
+		}
 
-			if innerTypesAreCompatible {
-				return validateValueAgainstType(fieldName, valueAsArray[0], compoundType.BaseType)
+		return false
+	case clickhouse.CompoundType:
+		if columnType.Name == "Array" {
```

Comment:
didn't we have constants for those clickhouse type names? I'd opt for using them in this PR.

Reply:
We don't have any existing constants for these ones (we only have constants for Elastic ones)

Reply:
bummer! yea I'm giving 👍 to whole PR then